Introduce Path, start using it all over the place.

hrydgard opened the pull request:
Prerequisite for upcoming changes in #14232 , will rebase that on this when merged.

The problem is that on Android content URIs, you can't simply append subdirectories to the filename, and cut the path in the normal ways. You need to decode, append, and re-encode - so all direct string manipulation of paths needs to go through a wrapper.

Still lots left to convert! Falling back on .ToString() and Path(std::string) to convert back and forth for now.

hrydgard committed:
Introduce Path, start using it all over the place.

Still lots left to convert!

Convert GetSysDirectory to return Path.

More buildfixing

Remove unnecessary Path( constructors

hrydgard committed:
Load IniFile from Path

Buildfixing

hrydgard committed:
Change parameter of DeleteDir and friends to Path

hrydgard committed:
Change Copy, Rename to use Path. Remove std::string version of Exists().

Buildfixes

Buildfix

hrydgard committed:
Use Path for fullName in dirlisting. Bugfixes.

Buildfixes

UWP: Buildfix.

headless: Buildfix.

Common: Buildfix.

iOS: Buildfixes.

libretro: Buildfix.

Qt: Buildfix.

hrydgard committed:
It never ends...

hrydgard committed:
Clean up PathBrowser a bit

Fix UWP build

Fix booting homebrew.

Qt buildfix

hrydgard commented:
Soon it will be worthwhile to merge and rebase #14232  on it and see if I can get alternate memstick directories to work.

Things I need to do first: 

- [ ] More testing!
- [ ] Remove std::*stream I/O from things like ini files.
- [ ] Fix Windows UWP ISO loading, broke somehow

hrydgard commented:
Section by section - what sections? It's, as you say, incredibly viral and there's a lot of expected and unexpected dependencies all across the code.

We can absolutely get smaller things like the != "" and OpenCPP file in first before the Great Pathening, but as for the typedef, I doubt it's worth the trouble...

So far most path errors show up quite easily in logs, too, once things break and are much easier to debug than random rendering  bugs. So think a bit of manual testing and then user feedback post-merge will get us through it. Fortunately we still have months to let it bubble if needed...

unknownbrackets commented on Core/Util/GameManager.cpp:
```diff
@@ -258,32 +260,37 @@ ZipFileContents DetectZipFileContents(struct zip *z, ZipFileInfo *info) {
+	Path dest = pspGame;
 	int error = 0;
+
+	// TODO(scoped): zip_open ain't gonna work. Might have to hack in one that can do it
```
Latest libzip seems to have a `zip_fdopen()`, so maybe we just need to update.

https://libzip.org/documentation/zip_fdopen.html

-[Unknown]

hrydgard replied:
Good catch!

hrydgard commented on Core/Debugger/SymbolMap.cpp:
```diff
@@ -65,15 +65,16 @@ void SymbolMap::Clear() {
 
 	std::lock_guard<std::recursive_mutex> guard(lock_);
 
+	// TODO(scoped): We're screwed here
```
Oh didn't see that. Yeah, true, reading the whole thing would be fine, too.

unknownbrackets reviewed (commented):
I know this is still a draft... added some small nits about dup Path, sorry.  Looks good, I suppose since \ vs / is only for wstrings, we shouldn't have any funky / problems with homebrew immediately.

-[Unknown]

hrydgard commented:
Thanks, lots of good feedback. I'll go through and apply it tonight.

hrydgard commented on Core/Config.cpp:
```diff
@@ -1614,27 +1614,27 @@ void Config::changeGameSpecific(const std::string &pGameId, const std::string &t
 		return false;
 	}
 
+	// TODO(scoped):
```
Just leftovers...

hrydgard commented on Common/File/FileUtil.cpp:
```diff
@@ -431,38 +496,33 @@ bool Copy(const std::string &srcFilename, const std::string &destFilename)
 
 	// buffer size
-#define BSIZE 1024
+#define BSIZE 4096
```
1024 is just ridiculously small these days.. Might as well do a full page.

hrydgard commented on UI/NativeApp.cpp:
```diff
@@ -1166,8 +1163,8 @@ void HandleGlobalMessage(const std::string &msg, const std::string &value) {
-			std::string dest = GetSysDirectory(DIRECTORY_SYSTEM) + (endsWithNoCase(value, ".jpg") ? "background.jpg" : "background.png");
-			File::Copy(value, dest);
+			Path dest = GetSysDirectory(DIRECTORY_SYSTEM) / (endsWithNoCase(value, ".jpg") ? "background.jpg" : "background.png");
+			File::Copy(Path(value), dest);
```
Actually not!

hrydgard commented on Common/File/Path.cpp:
```diff
@@ -0,0 +1,192 @@
+}
+
+void Path::operator /=(const std::string &subdir) {
+	path_ += path_ + "/" + subdir;
```
Oops! yeah, the this solution should work.

We don't use this operator much anyway.

hrydgard committed:
Address all the easy feedback. Some is left.

hrydgard commented:
Realized we are probably not supporting Windows network paths very well, so that will also need to be dealt with...

hrydgard committed:
Address additional feedback

hrydgard committed:
Remove unused function GetHostPath from filesystems

hrydgard commented on Core/Util/GameManager.cpp:
```diff
@@ -351,16 +357,16 @@ bool GameManager::DetectTexturePackDest(struct zip *z, int iniIndex, std::string
 		// Check for any supported game on their recent list and use that instead.
 		for (const std::string &path : g_Config.recentIsos) {
-			std::string recentID = GetGameID(path);
+			std::string recentID = GetGameID(Path(path));
```
Not yet!

hrydgard commented on Core/Loaders.cpp:
```diff
@@ -259,11 +251,12 @@ bool LoadFile(FileLoader **fileLoaderPtr, std::string *error_string) {
 				if (pos != std::string::npos) {
-					path = ResolvePBPDirectory(path);
-					pspFileSystem.SetStartingDirectory("ms0:/" + path.substr(pos));
+					dir = ResolvePBPDirectory(Path(dir)).ToString();
+					pspFileSystem.SetStartingDirectory("ms0:/" + dir.substr(pos));
```
It will have correct slashes because we use / slashes internally, and don't call .ToWString(). That said, there is a bit of an unholy mix of path types here.

unknownbrackets commented on Core/Util/GameManager.cpp:
```diff
@@ -516,21 +523,24 @@ bool GameManager::InstallMemstickGame(struct zip *z, const std::string &zipfile,
 			continue;
 		}
 		std::string outFilename = dest + zippedName.substr(info.stripChars);
+		// TODO(scoped): Tricky logic
```
Hmm, I guess not then...

-[Unknown]

hrydgard committed:
Additional feedback.

unknownbrackets committed:
GameManager: Base isDir off of zipped file name.

hrydgard commented:
Ha whoops, didn't mean to commit that half baked change

hrydgard committed:
UWP iso-loading fix

Having override:// as a fake file name was no good, Path would remove
one slash.

hrydgard committed:
Minor error message improvement

hrydgard commented:
Remaining issues I've found:

* Post shaders cannot be selected on Android

(more to be added as I test)

hrydgard committed:
Fix path bug installing homebrew

hrydgard committed:
Fix crash in homebrew store after installing game

hrydgard committed:
Fix ZipAssetReader file listing. Fixes post shaders on Android

hrydgard commented:
It's time to get this in, I've tested enough on enough platforms now.
